pilot/pkg/xds/ep_filters.go

```diff
@@ -61,73 +61,101 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocLbEndpointsAn
 			},
 		}
 
-		// Weight (number of endpoints) for the EDS cluster for each remote networks
-		remoteEps := map[network.ID]uint32{}
-		// Calculate remote network endpoints
+		// Create a map to keep track of the gateways used and their aggregate weights.
+		gatewayWeights := make(map[model.NetworkGateway]uint32)
+
+		// Process all of the endpoints.
 		for i, lbEp := range ep.llbEndpoints.LbEndpoints {
-			epNetwork := network.ID(istioMetadata(lbEp, "network"))
-			// This is a local endpoint or remote network endpoint
-			// but can be accessed directly from local network.
-			gatewaysForNetwork := b.push.NetworkManager().GatewaysForNetwork(epNetwork)
-			if identifier.IsSameOrEmpty(string(b.network), string(epNetwork)) || len(gatewaysForNetwork) == 0 {
-				// Copy on write.
-				clonedLbEp := proto.Clone(lbEp).(*endpoint.LbEndpoint)
-				clonedLbEp.LoadBalancingWeight = &wrappers.UInt32Value{
-					Value: uint32(multiples),
-				}
-				lbEndpoints.emplace(clonedLbEp, ep.tunnelMetadata[i])
-			} else {
-				if !b.canViewNetwork(epNetwork) {
-					continue
-				}
-				// cross-network traffic relies on mTLS to be enabled for SNI routing
-				// TODO BTS may allow us to work around this
-				if b.mtlsChecker.isMtlsDisabled(lbEp) {
-					continue
+			// Copy the endpoint in order to expand the load balancing weight.
+			// When multiplying, be careful to avoid overflow - clipping the
+			// result at the maximum value for uint32.
+			weight := b.scaleEndpointLBWeight(lbEp, scaleFactor)
+			lbEp := proto.Clone(lbEp).(*endpoint.LbEndpoint)
+			lbEp.LoadBalancingWeight = &wrappers.UInt32Value{
+				Value: weight,
+			}
+
+			istioEndpoint := ep.istioEndpoints[i]
+			epNetwork := istioEndpoint.Network
+			epCluster := istioEndpoint.Locality.ClusterID
+			gateways := b.selectNetworkGateways(epNetwork, epCluster)
```

Comment:
Are only cluster local endpoints visible to the cluster gateway? How did you make it?

Reply:
The gateway (like all proxies) will only see endpoints for the service that are cluster local or have been exported.